README.md
# aiVLE Worker

There are two parts in `aivle-worker` module:

- Sandbox: `sandbox.py` and everything under `profiles`, `requirements`, `scripts`
- (Celery) worker: everything else

If you want to test whether your aiVLE task bundle works, you may focus on *sandbox* part only.

## Requirement

This project is tested on the following environment. Similar environment may work but there is no guarantee.

* Ubuntu 20.04
* Firejail 0.9.62
* Python 3.8 (with `python3.8-venv`)

## Installation

From Test PyPI: `pip install -i https://test.pypi.org/simple/ aivle-worker`

From PyPI: `pip install aivle-worker`

## Getting Started

We assume you already installed `aivle-worker` by `pip install aivle-worker`.
(If `pip install` doesn't work, try `pip install -i https://test.pypi.org/simple/ aivle-worker` to fetch from Test PyPI)

1. Create `.env` (if you only intend to test sandbox, put dummy values on the RHS is enough):

Required fields:
```dotenv
API_BASE_URL=...
BROKER_URI=amqp://...# use sqs:// for AWS SQS, and remember to pip install celery[sqs] for dependencies
ACCESS_TOKEN=...

```
Optional fields (shown here are the default values):
```dotenv
TASK_QUEUE=gpu/private/default
CELERY_CONCURRENCY=1
WORKER_NAME=celery
ZMQ_PORT=15921
```

2. `python -m aivle-worker` **in the same directory** as your `.env` file (reason: `load_dotenv` will look for `.env`
in the current working directory).

## Test Sandbox

To test the standalone sandbox, you need to clone this repository and modify the source code under `src`.

### tl;dr

Under `src/aivle-worker/entry_point.py`, rewrite paths in `start_sandbox()` function, run that function.

**NOTE**:

1. There are three `/` in the URL (i.e. `file:///home/...`)
2. You need to use absolute path

### Details on how to run `start_sandbox()`

1. Rewrite paths in `start_sandbox()` function defined in `entry_point.py`
2. In the `if __name__ == "__main__":` block of `__main.py`, include `start_sandbox()` only
3. Make sure your current directoy is `src`, and `.env` is in `src` (i.e., `.env` is at the same level as 
`/aivle-worker`)
4. Run `python -m aivle-worker`

### How It Works

Referring to `settings.py`, by default the sandbox will load security profile from `profiles/aivle-base.profile`
and the shell script to create new virtual environment will be `scripts/create-venv.sh`, and the temporary folder will
be `/tmp/aivle-worker`. You don't need to change the settings if you're happy with the defaults.

The basic steps aivle Worker takes to evaluate an agent in your task bundle are:

1. Download and unzip task and agent archives
2. Create a new virtual environment (and activate it)
3. `pip install -r requirements.txt` (therefore in the root of your task bundle there must be a `requirements.txt`)
4. Inside the sandbox, run `bash bootstrap.sh` (similarly, `bootstrap.sh` comes from your task bundle)

## Test Sandbox + Worker (for devs only)

// TODO


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [0.1.1] - 2021-10-15

### Added

- Easily test run aiVLE task bundle along with sample agent using `start_sandbox` function

## [0.1.2] - 2021-12-18

### Added

- `CELERY_CONCURRENCY` option in `.env` for configurable concurrency
- Upgrade pip before `pip install` to avoid Git parsing error
- Print stderr for `create_venv` subroutine

### Fixed

- Changed `create-venv.sh` to avoid Anaconda Python (which has an incompatible impl of virtualenv)

## [0.1.3] - 2022-02-01

### Added

- Resource monitoring: CPU/GPU utlization, RAM/VRAM usage
- Resource-sensitive load balance: pause consuming tasks whenever resource constraint is violated, automatically resume
  when appropriate

## [0.1.4] - 2022-02-12

### Added

- Runtime error report
- RAM limit for individual tasks

## [0.2.0] - 2022-02-22

### Added

- VRAM limit for individual submission

## [0.2.1] - 2022-02-23

### Added

- Report error type upon submission and warden termination (currently reports RE, TLE, MLE, VLE)

## [0.2.2] - 2022-03-08

Packaging the module for publishing to PyPI

## [0.2.3] - 2022-03-10

### Added

- Command line interface - now you can use `aivle-worker` as a command after `pip install aivle-worker`

aivle-worker/apis.py
import json
import logging
import os

import requests

from errors import QueueInfoNotFound, StopConsumingError, ResumeConsumingError
from models import QueueInfo
from models import Submission, ExecutionOutput
from settings import API_BASE_URL, ACCESS_TOKEN, FULL_WORKER_NAME

logger = logging.getLogger("root")


def get_task_url(task_id: int):
    return API_BASE_URL + f"/tasks/{task_id}/download_grader/"


def get_agent_url(submission_id: int):
    return API_BASE_URL + f"/submissions/{submission_id}/download/"


def get_task_info(task_id: int):
    if task_id < 0:
        # return dummy data for negative task ID (local test)
        return {
            "id": 1,
            "grader": "courses/2/tasks/asdasd/grader/grader.zip",
            "template": "courses/2/tasks/asdasd/template/agent.zip",
            "daily_submission_limit": 100,
            "max_upload_size": 5120,
            "run_time_limit": 60,
            "course": 1,
            "name": "asdasd",
            "description": "asdasd",
            "ram_limit": 256,
            "vram_limit": 256,
            "opened_at": "2022-01-30T01:47:31+08:00",
            "deadline_at": "2022-01-30T01:47:30+08:00",
            "closed_at": "2022-01-30T01:47:30+08:00",
            "created_at": "2022-01-30T01:47:53.223015+08:00",
            "updated_at": "2022-02-07T22:03:11.244212+08:00",
            "eval_queue": 1
        }
    else:
        resp = requests.get(API_BASE_URL + f"/tasks/{task_id}/",
                            headers={"Authorization": f"Token {ACCESS_TOKEN}"})
        return resp.json()


def start_job(job_id, celery_task_id) -> Submission:
    worker_name = "unknown_worker"
    if os.getenv("WORKER_NAME") is not None:
        worker_name = os.getenv("WORKER_NAME")
    resp = requests.get(API_BASE_URL + f"/jobs/{job_id}/start_job/",
                        headers={"Authorization": f"Token {ACCESS_TOKEN}"},
                        data={
                            "worker_name": worker_name,
                            "task_id": celery_task_id
                        })
    if resp.status_code != 200:
        raise Exception(resp.content)
    obj = json.loads(resp.content)
    return Submission(sid=obj["submission"], task_url=get_task_url(obj["task"]),
                      agent_url=get_agent_url(obj["submission"]), task_id=int(obj["task"]))


ERROR_TIME_LIMIT_EXCEEDED = "TLE"
ERROR_MEMORY_LIMIT_EXCEEDED = "MLE"
ERROR_VRAM_LIMIT_EXCEEDED = "VLE"
ERROR_RUNTIME_ERROR = "RE"


def submit_job(job_id, task_id, output: ExecutionOutput):
    resp = requests.get(API_BASE_URL + f"/jobs/{job_id}/submit_job/",
                        headers={"Authorization": f"Token {ACCESS_TOKEN}"},
                        json={
                            "task_id": task_id,
                            "ok": output.ok,
                            "raw_log": output.raw,
                            "result": output.result,
                            "error": output.error,
                        })
    return resp


def update_job_error(job_id: int, task_id: str, error: str):
    resp = requests.get(API_BASE_URL + f"/jobs/{job_id}/update_job_error/",
                        headers={"Authorization": f"Token {ACCESS_TOKEN}"},
                        json={
                            "task_id": task_id,
                            "error": error
                        })
    return resp


def get_queue_info(queue_name: str) -> QueueInfo:
    resp = requests.get(API_BASE_URL + f"/queue/", params={"name": queue_name},
                        headers={"Authorization": f"Token {ACCESS_TOKEN}"})
    results = resp.json()["results"]
    if len(results) == 0:
        raise QueueInfoNotFound()
    result = results[0]
    return QueueInfo(pk=int(result["id"]), name=result["name"], cpu=result["cpu_required"], ram=result["ram_required"],
                     vram=result["vram_required"])


def stop_consuming(queue: QueueInfo):
    resp = requests.get(API_BASE_URL + f"/queue/{queue.pk}/stop_consuming/", params={"worker": FULL_WORKER_NAME},
                        headers={"Authorization": f"Token {ACCESS_TOKEN}"})
    if resp.status_code != 200:
        raise StopConsumingError(f"stop consuming failed [{resp.status_code}]: {resp.content}")


def resume_consuming(queue: QueueInfo):
    resp = requests.get(API_BASE_URL + f"/queue/{queue.pk}/resume_consuming/", params={"worker": FULL_WORKER_NAME},
                        headers={"Authorization": f"Token {ACCESS_TOKEN}"})
    if resp.status_code != 200:
        raise ResumeConsumingError(f"resume consuming failed [{resp.status_code}]: {resp.content}")
